nixos/modules/services/network-filesystems/ipfs.nix

```diff
@@ -0,0 +1,114 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  inherit (pkgs) ipfs;
+
+  cfg = config.services.ipfs;
+
+
+  ipfsFlags = ''${if cfg.autoMigrate then "--migrate" else ""} ${if cfg.enableGC then "--enable-gc" else ""} ${toString cfg.extraFlags}'';
+
+in
+
+{
+
+  ###### interface
+
+  options = {
+
+    services.ipfs = {
+
+      enable = mkOption {
+        default = false;
+        description = ''
+          Whether to start a daemon for the Interplanetary
+          file system (IPFS).
+        '';
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "ipfs";
+        description = "User under which the IPFS daemon runs";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "ipfs";
+        description = "Group under which the IPFS daemon runs";
+      };
+
+      dataDir = mkOption {
+        type = types.str;
+        default = "/var/lib/ipfs";
+        description = "The data dir for IPFS";
+      };
+
+      autoMigrate = mkOption {
+        default = false;
+        description = ''
+          Whether IPFS should try to migrate the file system automatically.
+        '';
+      };
+
+      enableGC = mkOption {
+        default = false;
+        description = ''
+          Whether to enable automatic garbadge collection.
+        '';
+      };
+
+      extraFlags = mkOption {
+        type = types.listOf types.str;
+        description = "Extra flags passed to the IPFS daemon";
+        default = [];
+      };
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.ipfs ];
+
+    users.extraUsers = optionalAttrs (cfg.user == "ipfs") singleton {
+      name = "ipfs";
```

Comment:
nit: Use `mkIf` instead of `optionalAttrs`.
nit: `singleton` will create a list element which is not extensible, use the user name instead, in which case the name would not be needed.

```nix
{
  users.extraUsers = mkIf (cfg.user == "ipfs") {
    ipfs = {
       group = cfg.group;
       home = …;
       …;
    };
  };
}
```